src/bin/report_collector.rs

```diff
@@ -0,0 +1,383 @@
+    ff::{FieldType, Fp31, Fp32BitPrime},
+    helpers::query::{IpaQueryConfig, QueryConfig, QueryType},
+    hpke::{KeyRegistry, PublicKeyOnly},
+    net::{ClientIdentity, MpcHelperClient},
+    protocol::{BreakdownKey, MatchKey},
+    report::{KeyIdentifier, DEFAULT_KEY_ID},
+    test_fixture::{
+        ipa::{ipa_in_the_clear, IpaSecurityModel, TestRawDataRecord},
+        EventGenerator, EventGeneratorConfig,
+    },
+};
+use rand::thread_rng;
+use std::{
+    error::Error,
+    fmt::Debug,
+    fs,
+    fs::OpenOptions,
+    io,
+    io::{stdout, Write},
+    path::PathBuf,
+    time::{Duration, Instant},
+};
+use tokio::time::sleep;
+
+#[derive(Debug, Parser)]
+#[clap(
+    name = "mpc-client",
+    about = "CLI to execute test scenarios on IPA MPC helpers"
+)]
+#[command(about)]
+struct Args {
+    #[clap(flatten)]
+    logging: Verbosity,
+
+    /// Path to helper network configuration file
+    #[arg(long)]
+    network: Option<PathBuf>,
+
+    /// Use insecure HTTP
+    #[arg(short = 'k', long)]
+    disable_https: bool,
+
+    /// Seconds to wait for server to be running
+    #[arg(short, long, default_value_t = 0)]
+    wait: usize,
+
+    #[clap(flatten)]
+    input: CommandInput,
+
+    #[command(subcommand)]
+    action: TestAction,
+}
+
+#[derive(Debug, Parser)]
+pub struct CommandInput {
+    #[arg(
+        long,
+        help = "Read the input from the provided file, instead of standard input"
+    )]
+    input_file: Option<PathBuf>,
+
+    #[arg(value_enum, long, default_value_t = FieldType::Fp32BitPrime, help = "Convert the input into the given field before sending to helpers")]
+    field: FieldType,
+}
+
+impl From<&CommandInput> for InputSource {
+    fn from(source: &CommandInput) -> Self {
+        if let Some(ref file_name) = source.input_file {
+            InputSource::from_file(file_name)
+        } else {
+            InputSource::from_stdin()
+        }
+    }
+}
+
+#[derive(Debug, Subcommand)]
+enum TestAction {
+    /// Execute IPA in semi-honest honest majority setting
+    SemiHonestIpa(IpaQueryConfig),
+    /// Execute IPA in malicious honest majority setting
+    MaliciousIpa(IpaQueryConfig),
+    /// Generate inputs for IPA
+    GenIpaInputs {
+        /// Number of records to generate
+        #[clap(long, short = 'n')]
+        count: u32,
+
+        /// The destination file for generated records
+        #[arg(long)]
+        output_file: Option<PathBuf>,
+
+        #[clap(flatten)]
+        gen_args: EventGeneratorConfig,
+    },
+}
+
+#[derive(Debug, clap::Args)]
+struct GenInputArgs {
+    /// Maximum records per user
+    #[clap(long)]
+    max_per_user: u32,
+    /// number of breakdowns
+    breakdowns: u32,
+}
+
+async fn clients_ready(clients: &[MpcHelperClient; 3]) -> bool {
+    clients[0].echo("").await.is_ok()
+        && clients[1].echo("").await.is_ok()
+        && clients[2].echo("").await.is_ok()
+}
+
+fn validate<I, S>(expected: I, actual: I)
```

Comment:
can we move it to `playbook` module and re-use in both test_mpc and rc binaries?

Reply:
n00b qn: What is the purpose of playbook? What kind of code should go there?